sql/postgres.up.psql

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2018 robzon.
+ * See the LICENSE file for more information.
+ *
+ * Notes:
+ *
+ * As per https://tools.ietf.org/html/rfc6122#page-4
+ *
+ * - Username MUST NOT be zero bytes in length and MUST NOT be more than 1023 bytes in length
+ * - JIDs total length cannot be more than 3071 bytes
+ *
+ */
+
+-- Functions to manage updated_at timestamps
+
+CREATE OR REPLACE FUNCTION enable_updated_at(_tbl regclass) RETURNS VOID AS $$
+BEGIN
+    EXECUTE format('CREATE TRIGGER set_updated_at BEFORE UPDATE ON %s
+                    FOR EACH ROW EXECUTE PROCEDURE set_updated_at()', _tbl);
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE FUNCTION set_updated_at() RETURNS trigger AS $$
+BEGIN
+    IF (
+        NEW IS DISTINCT FROM OLD AND
+        NEW.updated_at IS NOT DISTINCT FROM OLD.updated_at
+    ) THEN
+        NEW.updated_at := current_timestamp;
+    END IF;
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+-- users
+
+CREATE TABLE IF NOT EXISTS users (
+    username            VARCHAR(1023) PRIMARY KEY,
+    password            TEXT NOT NULL,
+    last_presence       TEXT NOT NULL,
+    last_presence_at    TIMESTAMP WITH TIME ZONE NOT NULL,
+    updated_at          TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
+    created_at          TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW()
+);
+
+SELECT enable_updated_at('users');
+
+-- roster_notifications
+
+CREATE TABLE IF NOT EXISTS roster_notifications (
+    contact     VARCHAR(1023) NOT NULL,
+    jid         TEXT NOT NULL,
+    elements    TEXT NOT NULL,
+    updated_at  TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
+    created_at  TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
+
+    PRIMARY KEY (contact, jid)
+);
+
+CREATE INDEX IF NOT EXISTS i_roster_notifications_jid ON roster_notifications(jid);
```

Comment:
Could be, I want to add benchmarks later on, that will make tweaking performance stuff much easier!

Reply:
Maybe we should avoid the premature optimization then?

Reply:
Agreed. To be honest I left it here only because it was in the mysql script and I don't know enough about postgres to make that call myself. I'm all for removing it if it's not necessary at this point. Will do it in a little bit.

Reply:
Removed all the indexes that were a subset of a different index.